File: groupware/packages/net-utils/src/http-client.ts
/**
 * 서버 간 HTTP 클라이언트 (fetch 기반, 의존성 0)
 * 계약: docs/specs/service-http-contract.md §3.
 *  - X-Service-Token 자동 주입 / 타임아웃 / 재시도(네트워크, 타임아웃, 5xx) / 정규화 에러(HttpError)
 */
import {
  HttpError,
  httpErrorFromNetwork,
  httpErrorFromResponse,
} from './http-error';
import { correlationHeaders } from './request-context';

export interface HttpClientOptions {
  // 베이스 URL (예: http://video-model:8000)
  baseUrl: string;
  // 호출마다 X-Service-Token 을 생성해 주입 (없으면 미주입)
  serviceToken?: () => string;
  // 모든 요청에 추가할 기본 헤더
  defaultHeaders?: Record<string, string>;
  // 요청 타임아웃(ms). 기본 10000.
  timeoutMs?: number;
  // 재시도 횟수(네트워크/타임아웃/5xx 한정). 기본 0.
  retries?: number;
  // 재시도 대기(ms). 기본 min(2^n, 30)*1000.
  retryDelayMs?: (attempt: number) => number;
  // 테스트/대체용 fetch 주입. 기본 global fetch.
  fetchImpl?: typeof fetch;
}

export interface HttpClient {
  get<T>(path: string, init?: RequestInit): Promise<T>;
  post<T>(path: string, body?: unknown, init?: RequestInit): Promise<T>;
  patch<T>(path: string, body?: unknown, init?: RequestInit): Promise<T>;
  delete<T>(path: string, body?: unknown, init?: RequestInit): Promise<T>;
}

const defaultDelay = (attempt: number) => Math.min(2 ** attempt, 30) * 1000;
const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

export function createHttpClient(options: HttpClientOptions): HttpClient {
  const {
    baseUrl,
    serviceToken,
    defaultHeaders,
    timeoutMs = 10_000,
    retries = 0,
    retryDelayMs = defaultDelay,
    fetchImpl = fetch,
  } = options;

  async function once(
    method: string,
    path: string,
    body: unknown,
    init?: RequestInit,
  ): Promise<unknown> {
    // 상관관계 헤더를 먼저 깔고 defaultHeaders 가 덮을 수 있게 둔다(호출부 오버라이드 허용)
    // 컨텍스트가 없으면(워커/크론) 빈 객체라 비용이 사실상 0이다.
    const headers: Record<string, string> = { ...correlationHeaders(), ...defaultHeaders };
    if (body !== undefined) headers['Content-Type'] = 'application/json';
    if (serviceToken) headers['X-Service-Token'] = serviceToken();

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeoutMs);
    let res: Response;
    try {
      res = await fetchImpl(`${baseUrl}${path}`, {
        ...init,
        method,
        headers: { ...headers, ...(init?.headers as Record<string, string>) },
        body: body !== undefined ? JSON.stringify(body) : undefined,
        signal: controller.signal,
      });
    } catch (err) {
      // 네트워크 실패/타임아웃(abort) → 503 (재시도 대상)
      throw httpErrorFromNetwork(
        controller.signal.aborted ? '서버 응답 시간이 초과되었습니다' : '서비스에 연결할 수 없습니다',
      );
    } finally {
      clearTimeout(timer);
    }

    const text = await res.text();
    const data = text ? safeJson(text) : undefined;
    if (!res.ok) throw httpErrorFromResponse(res.status, data);
    return data;
  }

  async function request<T>(
    method: string,
    path: string,
    body: unknown,
    init?: RequestInit,
  ): Promise<T> {
    let lastErr: unknown;
    for (let attempt = 0; attempt <= retries; attempt++) {
      try {
        return (await once(method, path, body, init)) as T;
      } catch (err) {
        lastErr = err;
        if (attempt < retries && isRetryable(err)) {
          await sleep(retryDelayMs(attempt));
          continue;
        }
        throw err;
      }
    }
    throw lastErr;
  }

  return {
    get: <T>(path: string, init?: RequestInit) => request<T>('GET', path, undefined, init),
    post: <T>(path: string, body?: unknown, init?: RequestInit) =>
      request<T>('POST', path, body, init),
    patch: <T>(path: string, body?: unknown, init?: RequestInit) =>
      request<T>('PATCH', path, body, init),
    delete: <T>(path: string, body?: unknown, init?: RequestInit) =>
      request<T>('DELETE', path, body, init),
  };
}

/** 네트워크/타임아웃(503) 또는 5xx 만 재시도. 4xx 는 재시도 안 함 */
function isRetryable(err: unknown): boolean {
  return err instanceof HttpError && err.statusCode >= 500;
}

function safeJson(text: string): unknown {
  try {
    return JSON.parse(text);
  } catch {
    return text;
  }
}

File: groupware/apps/api/fastapi/net-utils/src/csc_net_utils/http_client.py
"""서버 간 아웃바운드 HTTP 클라이언트 (httpx async).

`csc-net-utils[http]` 익스트라에서만 사용 가능(httpx 필요). httpx 미설치 앱에서 import 되지 않도록
패키지 `__init__` 은 이 모듈을 노출하지 않는다. 필요한 앱이 명시적으로 import 한다:
    from csc_net_utils.http_client import ServiceHttpClient

계약: docs/specs/service-http-contract.md §3 (토큰 주입, 타임아웃, 재시도, 에러 봉투).
"""

from __future__ import annotations

import asyncio
from collections.abc import Callable
from typing import Any

import httpx

from .request_context import correlation_headers


class ServiceHttpError(Exception):
    """정규화된 서버 간 HTTP 에러(상태, 봉투 보존)."""

    def __init__(self, status_code: int, message: str, body: Any | None = None) -> None:
        super().__init__(message)
        self.status_code = status_code
        self.body = body


def _extract_message(body: Any, fallback: str) -> str:
    if isinstance(body, dict):
        for key in ("error", "message", "detail"):
            value = body.get(key)
            if isinstance(value, str) and value:
                return value
    return fallback


class ServiceHttpClient:
    """X-Service-Token 자동 주입 + 타임아웃 + 재시도(네트워크/타임아웃/5xx)."""

    def __init__(
        self,
        base_url: str,
        *,
        token_provider: Callable[[], str] | None = None,
        timeout: float = 10.0,
        retries: int = 0,
        backoff: Callable[[int], float] | None = None,
        transport: httpx.AsyncBaseTransport | None = None,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._token = token_provider
        self._timeout = timeout
        self._retries = retries
        self._backoff = backoff or (lambda attempt: float(min(2**attempt, 30)))
        self._transport = transport  # 테스트 주입용(MockTransport). prod 는 None.
        # 커넥션 재사용: 아래 _client() 주석 참고.
        self._pool: httpx.AsyncClient | None = None
        self._pool_lock = asyncio.Lock()

    def _client(self) -> httpx.AsyncClient:
        """지연 생성되는 공유 AsyncClient.

        요청마다 `async with httpx.AsyncClient(...)` 로 새로 만들면 호출마다 커넥션 풀을 만들고
        버려서 TCP 와 TLS 핸드셰이크가 매번 발생한다. 트래픽이 늘수록 지연과 TIME_WAIT 소켓이
        함께 늘어난다. 클라이언트를 인스턴스 수명 동안 재사용하면 keep-alive 로 커넥션이 유지된다.

        어댑터들이 이 클라이언트를 싱글톤으로 들고 있으므로(video-model worker 등) 수명은 그들과 같다.
        """
        if self._pool is None:
            self._pool = httpx.AsyncClient(
                timeout=self._timeout,
                transport=self._transport,
                # 서버 간 호출은 상대가 소수(도커 서비스명)라 작은 풀로 충분하다.
                limits=httpx.Limits(max_keepalive_connections=20, max_connections=100),
            )
        return self._pool

    async def aclose(self) -> None:
        """공유 클라이언트를 닫는다(앱 종료 시). 호출하지 않아도 프로세스 종료로 정리된다."""
        async with self._pool_lock:
            if self._pool is not None:
                await self._pool.aclose()
                self._pool = None

    def _headers(self, has_body: bool) -> dict[str, str]:
        # 상관관계(trace/request id) 전파: 컨텍스트가 없으면 빈 dict 라 비용이 사실상 0.
        headers: dict[str, str] = correlation_headers()
        if has_body:
            headers["Content-Type"] = "application/json"
        if self._token is not None:
            headers["X-Service-Token"] = self._token()
        return headers

    async def request(
        self, method: str, path: str, *, json: Any | None = None
    ) -> Any:
        last_exc: Exception | None = None
        for attempt in range(self._retries + 1):
            try:
                resp = await self._client().request(
                    method,
                    f"{self._base_url}{path}",
                    json=json,
                    headers=self._headers(json is not None),
                )
                if resp.status_code >= 400:
                    body = _safe_json(resp)
                    err = ServiceHttpError(
                        resp.status_code,
                        _extract_message(body, "서버 간 요청 실패"),
                        body,
                    )
                    # 4xx 는 재시도하지 않음.
                    if resp.status_code < 500 or attempt >= self._retries:
                        raise err
                    last_exc = err
                else:
                    return _safe_json(resp)
            except (httpx.TransportError, httpx.TimeoutException) as exc:
                last_exc = exc
                if attempt >= self._retries:
                    raise ServiceHttpError(503, "서비스에 연결할 수 없습니다") from exc
            await asyncio.sleep(self._backoff(attempt))
        assert last_exc is not None
        raise last_exc

    async def get(self, path: str) -> Any:
        return await self.request("GET", path)

    async def post(self, path: str, json: Any | None = None) -> Any:
        return await self.request("POST", path, json=json)

    async def patch(self, path: str, json: Any | None = None) -> Any:
        return await self.request("PATCH", path, json=json)

    async def delete(self, path: str) -> Any:
        return await self.request("DELETE", path)


def _safe_json(resp: httpx.Response) -> Any:
    try:
        return resp.json()
    except ValueError:
        return resp.text or None
